src/ambulance_game/markov/waiting.py

```diff
@@ -0,0 +1,340 @@
+import numpy as np
+import functools
+
+from .utils import (
+    is_waiting_state,
+    is_accepting_state,
+    expected_time_in_markov_state_ignoring_arrivals,
+)
+
+from .markov import (
+    build_states,
+    get_transition_matrix,
+    get_steady_state_algebraically,
+    get_markov_state_probabilities,
+)
+
+
+@functools.lru_cache(maxsize=None)
+def get_recursive_waiting_time(
+    state,
+    patient_type,
+    lambda_a,
+    lambda_o,
+    mu,
+    num_of_servers,
+    threshold,
+    system_capacity,
+    parking_capacity,
+):
+    """Performs a recursive algorithm to get the expected waiting time of patients
+    when they enter the model at a given state. Given an arriving state the algorithm
+    moves down to all subsequent states until it reaches one that is not a waiting
+    state.
+
+    Others:
+        - If (u,v) not a waiting state: return 0
+        - Arriving state s_a = (u, v + 1)
+        - Next state s_d = (0, v - 1)
+        - w(u,v) = c(s_a) + w(s_d)
+
+    Ambulance:
+        - If (u,v) not a waiting state: return 0
+        - Next state:       s_n = (u-1, v),    if u >=1 and v=T
+                            s_n = (u, v - 1),  otherwise
+        - w(u,v) = c(u,v) + w(s_n)
+
+    Note:   For all "others" patients the recursive formula acts in a linear manner
+    meaning that an individual will have the same waiting time when arriving at
+    either state of the same column e.g (2, 3) or (5, 3).
+
+    Parameters
+    ----------
+    state : tuple
+    patient_type : string
+    lambda_a : float
+    lambda_o : float
+    mu : float
+    num_of_servers : int
+    threshold : int
+    system_capacity : int
+    parking_capacity : int
+
+    Returns
+    -------
+    float
+        The expected waiting time from the arriving state of an individual until
+        service
+    """
+    if not is_waiting_state(state, num_of_servers):
+        return 0
+    if state[0] >= 1 and state[1] == threshold:
+        next_state = (state[0] - 1, state[1])
+    else:
+        next_state = (state[0], state[1] - 1)
+
+    wait = expected_time_in_markov_state_ignoring_arrivals(
+        state, patient_type, num_of_servers, mu, threshold
+    )
+    wait += get_recursive_waiting_time(
+        next_state,
+        patient_type,
+        lambda_a,
+        lambda_o,
+        mu,
+        num_of_servers,
+        threshold,
+        system_capacity,
+        parking_capacity,
+    )
+    return wait
+
+
+def mean_waiting_time_formula(
+    all_states,
+    pi,
+    patient_type,
+    lambda_a,
+    lambda_o,
+    mu,
+    num_of_servers,
+    threshold,
+    system_capacity,
+    parking_capacity,
+    formula="closed_form",
+):
+    """Get the mean waiting time by using the recursive formula or a closed-form
+    formula
+
+    Recursive Formula:
```

Comment:
Is this recursive?

Reply:
Nope. That shouldn't have said `Recursive Formula`